karabo_data/reader.py

```diff
@@ -662,6 +662,51 @@ def get_array(self, source, key, extra_dims=None, roi=by_index[...]):
             sorted(non_empty, key=lambda a: a.coords['trainId'][0]), dim='trainId'
         )
 
+    def get_dask_array(self, source, key):
+        """Get a Dask array for the specified data field.
+
+        Dask is a system for lazy parallel computation. This method doesn't
+        actually load the data, but gives you an array-like object which you
+        can operate on. Dask loads the data and calculates results when you ask
+        it to, e.g. by calling a ``.compute()`` method.
+        See the Dask documentation for more details.
+
+        If your computation depends on reading lots of data, consider creating
+        a dask.distributed.Client before calling this.
+        If you don't do this, Dask uses threads by default, which is not
+        efficient for reading HDF5 files.
+
+        Parameters
+        ----------
+        source: str
+            Source name, e.g. "SPB_DET_AGIPD1M-1/DET/7CH0:xtdf"
+        key: str
+            Key of parameter within that device, e.g. "image.data".
+        """
+        import dask.array as da
+        chunks = sorted(
+            self._find_data_chunks(source, key),
+            key=lambda x: x.train_ids[0] if x.train_ids.size else 0,
+        )
+
+        chunks_darrs = []
+        for chunk in chunks:
+            chunk_dim0 = np.sum(chunk.counts)
+            chunk_shape = (chunk_dim0,) + chunk.dataset.shape[1:]
+
+            # Find chunk size of maximum 1 GB
+            # TODO: optional user control of chunking
+            limit = 1024 ** 3
+            while np.product(chunk_shape) > limit and chunk_dim0 > 1:
```

Comment:
This is maybe just me being a slow thinker but you could add a comment here telling what you're trying to achieve.

Reply:
Is "Find chunk size of maximum 1 GB" not it?

In fact, that's not exactly what it's doing, because I don't take into account the size of each entry, so it's 1 billion array elements. I should probably fix that.

Reply:
Basically arbitrary. I did briefly play with making the chunks bigger and smaller ([performance testing log](https://git.xfel.eu/gitlab/dataAnalysis/data-reading-performance/blob/master/results.txt)), and maybe bigger was slightly better, but the timings vary so much that it's really hard to evaluate changes that don't make a massive difference.